### PR TITLE
Swift: Query for REDOS (Regular Expression Denial Of Service)

### PR DESCRIPTION
**Pull request by geoffw0**
Adds a query for REDOS (Regular Expression Denial Of Service), built atop the regular expressions library from https://github.com/github/codeql/pull/13470.

At present there are significant limitations (the largest of which being that the regex library doesn't support regular expression literals yet), but I believe the query adds value in its current state, and will find more results as we improve the library.

TODO:
- [x] team review
- [x] doc review
- [x] DCA run

### CODE REVIEW

File: swift/ql/test/query-tests/Security/CWE-1333/ReDoS.swift
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-1333/ReDoS.swift:85]: There are many more test cases for the core logic of this query, i.e. identifying REDOS vulnerable regular expressions, in the library test file https://github.com/github/codeql/blob/main/swift/ql/test/library-tests/regex/redos_variants.swift .  These include realistic cases and edge cases designed to test the limits of the engine.

Comment (d10c) [reply swift/ql/test/query-tests/Security/CWE-1333/ReDoS.swift:85]: Sounds good. As long as this query delegates to the same implementation, there's no need to copy those test cases here.

**Review by d10c — COMMENTED** [review]
LGTM with one minor suggestion. Will this be run as part of the standard query suite? If so, might want to do a DCA run before merging.

**Comment by MathiasVP** [comment]
> LGTM with one minor suggestion. Will this be run as part of the standard query suite? If so, might want to do a DCA run before merging.

We should definitely have a DCA run for this. It'll go directly into the Code Scanning suite as it:
- Is tagged as a `security` query, and
- is tagged with `@precision high`.

File: swift/ql/test/query-tests/Security/CWE-1333/ReDoS.swift
Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-1333/ReDoS.swift:85]: Yep, the query tests mostly just confirm that everything is wired up correctly and presents the results in a reasonable manner.

File: swift/ql/src/queries/Security/CWE-1333/ReDoSIntroduction.inc.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-1333/ReDoSIntroduction.inc.qhelp]: Ah, well spotted, updated.

**Comment by geoffw0** [comment]
Thanks for reviewing...

> Will this be run as part of the standard query suite?

As @mathiasvp said this will run in the standard suite due to its metadata.  We sometimes start queries at a lower `@precision` but this tends to be when there are known and common false positives, and I've not seen that.

> If so, might want to do a DCA run before merging.

Definitely.  I've just started a run.

> This library also frequently appear as the cause of various timeouts in other languages 😄.

Yeah, we've already encountered (and fixed) a timeout in one of the test regexps in the library PR (https://github.com/github/codeql/pull/13470/).  I didn't encounter any problems on the MRVA top 1000 though.  In any case I'm _moderately_ confident any further problems we encounter will be isolated to specific forms of regex (ironically, given what this query is about) and can be investigated and solved.

**Comment by geoffw0** [comment]
DCA LGTM.  Analysis time is not measurably slower, which if anything is a slight surprise as this is the first query to use the new regex library.

**Review by d10c — APPROVED** [review]
LGTM! :shipit:

**Comment by geoffw0** [comment]
@github/docs-content-codeql please could you confirm this one hasn't been forgotten?

**Comment by geoffw0** [comment]
Note that the qhelp preview generated all sorts of stuff, only the Swift `.qhelp` is under review here.

File: swift/ql/src/queries/Security/CWE-1333/ReDoSIntroduction.inc.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-1333/ReDoSIntroduction.inc.qhelp:34]:
```suggestion
      time complexity does not matter.
```

**Review by subatoi — DISMISSED** [review]
Looks good 👍 —just had some very minor Docs suggestions

**Comment by geoffw0** [comment]
Thanks @subatoi, all suggestions accepted!

Now this just needs a re-approval from someone and merging.

**Review by subatoi — APPROVED** [review]

**Pull request merged**
